pkg/operator/staticpod/controller/guard/guard_controller.go

```diff
@@ -0,0 +1,252 @@
+package guard
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/ghodss/yaml"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	appsv1listers "k8s.io/client-go/listers/apps/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+)
+
+const (
+	infrastructureClusterName = "cluster"
+	clusterConfigName         = "cluster-config-v1"
+	clusterConfigKey          = "install-config"
+	clusterConfigNamespace    = "kube-system"
+)
+
+type replicaCountDecoder struct {
+	ControlPlane struct {
+		Replicas string `yaml:"replicas,omitempty"`
+	} `yaml:"controlPlane,omitempty"`
+}
+
+// GuardController ensures that the guarding workload and pdb are available and running
+type GuardController struct {
+	operatorClient       v1helpers.OperatorClient
+	kubeClient           kubernetes.Interface
+	deploymentLister     appsv1listers.DeploymentLister
+	nodeLister           corev1listers.NodeLister
+	infrastructureLister configv1listers.InfrastructureLister
+	clusterTopology      configv1.TopologyMode
+	replicaCount         int
+	targetNamespace      string
+	deploymentManifest   []byte
+	pdbManifest          []byte
+	cliImagePullSpec     string
+	factory              *factory.Factory
+	eventRecorder        events.Recorder
+}
+
+var _ factory.Controller = &GuardController{}
+
+func NewController(operatorClient v1helpers.OperatorClient,
```

Comment:
We really want just a generic version of this, right? https://github.com/openshift/library-go/blob/master/pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go

you want to stub https://github.com/openshift/library-go/blob/master/pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go#L160 and you want to provide a hook https://github.com/openshift/library-go/blob/master/pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go#L163